### ISSUE TITLE
Compatible Django/Python versions are outdated

### ISSUE THREAD

**Issue by jsma** [django-oscar/django-oscar#3908]
Oscar maintains compatibility with versions of Python and Django that have already reached EOL but does not support the most recent versions. Apologies in advance if I've missed an issue or other PR where this is already under discussion or active development.

### Issue Summary

According to the release notes for Oscar 3.1:

https://github.com/django-oscar/django-oscar/blob/fe37a51c040303b6f0251d1973d39694c3eaba88/docs/source/releases/v3.1.rst?plain=1#L14-L19

According to the release notes for the still in-development Oscar 3.2:

https://github.com/django-oscar/django-oscar/blob/fe37a51c040303b6f0251d1973d39694c3eaba88/docs/source/releases/v3.2.rst?plain=1#L14-L19

I see that there is an [in progress PR](https://github.com/django-oscar/django-oscar/pull/3864) to update the release notes but none of the proposed changes modifies the Django/Python support section.

The currently supported upstream versions as of April 2022 are:
- Python 3.7 to 3.10
- Django 3.2 to 4.0

Details on EOL versions and latest releases:
- Extended support for Django 2.2 ends this month, April 2022
- Extended support for Django 3.1 ended on December 7, 2021
- Django 4.0 was released on December 7, 2021
- Python 3.6 was EOL as of December 2021
- Python 3.10 was released on October 4, 2021 and will be included in the upcoming Ubuntu 22.04 LTS release later this month.

Oscar 3.1 pins Django at:

https://github.com/django-oscar/django-oscar/blob/04dd391c900537a61dbb0ae5250ca5c2df6bbc4b/setup.py#L21

The in-progress Oscar 3.2 pins Django at:

https://github.com/django-oscar/django-oscar/blob/fe37a51c040303b6f0251d1973d39694c3eaba88/setup.py#L21

At the moment, it's not possible to upgrade a project to Django 4.0.x given Oscar's constraints.
I do not know yet if there are any actual incompatibilities with Python 3.10.x.

Are the maintainers of this project open to pull requests to:

1. Remove references to EOL versions in Oscar 3.2
2. Add Django 4.0/Python 3.10 compatibility to Oscar 3.2
 
Is there a target date of when Oscar 3.2 will be released? If not or if it is still many months in the future, would the maintainers be open to a patch release of Oscar 3.1 that adds support for these versions and removes compatibility with EOL versions, which could then be merged into the in-development version?

Thanks!

**Comment by solarissmoke** [comment]
Thanks for your detailed report. We are aware that support for Django 4 is missing - unfortunately this is currently blocked by some of the packages that Oscar depends on (e.g., when I checked a couple of weeks ago django-haystack was not compatible with Django 4). A PR to add support for Django 4 is welcome - but likely still blocked by some packages. It is likely that we'll have to wait a bit - in the mean time projects are encouraged to use Django 3.2 LTS.

We will drop support for EOL versions of Django in the next release. We expect to cut a new release this month.

**Comment by jsma** [comment]
I see. We don't use haystack and actually have to do this in settings so Oscar won't complain:

```
HAYSTACK_CONNECTIONS = {"default": {}}
```

Oscar has always been a bit overkill for our needs, so given the chain of dependencies holding us back, it might be a better use of our time to replace Oscar with a simpler home grown solution. 

Thanks for the reply! Feel free to close.

**Comment by jayvdb** [comment]
haystack has done a release with the Django 4 compatibility.

oscar's tests failure on Django 4.1.

**Comment by jayvdb** [comment]
To avoid conflicts, https://github.com/django-oscar/django-oscar/pull/3938 should be merged first.

**Comment by jayvdb** [comment]
ftr, Django 4.1 causes 95 failures, and from a quick scan they are mostly like these two.  Probably one root cause of them all.
```py
[  327s] ________________ TestScales.test_returns_zero_for_empty_basket _________________
[  327s] 
[  327s] self = <tests.integration.shipping.test_scales.TestScales testMethod=test_returns_zero_for_empty_basket>
[  327s] 
[  327s]     def test_returns_zero_for_empty_basket(self):
[  327s]         basket = Basket()
[  327s]     
[  327s]         scale = Scale(attribute_code='weight')
[  327s] >       self.assertEqual(0, scale.weigh_basket(basket))
[  327s] 
[  327s] /home/abuild/rpmbuild/BUILD/django-oscar-3.1/tests/integration/shipping/test_scales.py:32: 
[  327s] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[  327s] /home/abuild/rpmbuild/BUILDROOT/python-django-oscar-3.1-8.8.x86_64/usr/lib/python3.8/site-packages/oscar/apps/shipping/scales.py:33: in weigh_basket
[  327s]     for line in basket.lines.all():
[  327s] /usr/lib/python3.8/site-packages/django/db/models/manager.py:164: in all
[  327s]     return self.get_queryset()
[  327s] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[  327s] 
[  327s] self = <django.db.models.fields.related_descriptors.create_reverse_many_to_one_manager.<locals>.RelatedManager object at 0x7f385a64f880>
[  327s] 
[  327s]     def get_queryset(self):
[  327s]         # Even if this relation is not to pk, we require still pk value.
[  327s]         # The wish is that the instance has been already saved to DB,
[  327s]         # although having a pk value isn't a guarantee of that.
[  327s]         if self.instance.pk is None:
[  327s] >           raise ValueError(
[  327s]                 f"{self.instance.__class__.__name__!r} instance needs to have a "
[  327s]                 f"primary key value before this relationship can be used."
[  327s]             )
[  327s] E           ValueError: 'Basket' instance needs to have a primary key value before this relationship can be used.
[  327s] 
[  327s] /usr/lib/python3.8/site-packages/django/db/models/fields/related_descriptors.py:687: ValueError
[  327s] _______________ TestScales.test_weighs_uses_specified_attribute ________________
[  327s] 
[  327s] self = <tests.integration.shipping.test_scales.TestScales testMethod=test_weighs_uses_specified_attribute>
[  327s] 
[  327s]     def test_weighs_uses_specified_attribute(self):
[  327s]         scale = Scale(attribute_code='weight')
[  327s] >       p = factories.create_product(attributes={'weight': '1'})
[  327s] 
[  327s] /home/abuild/rpmbuild/BUILD/django-oscar-3.1/tests/integration/shipping/test_scales.py:14: 
[  327s] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[  327s] /home/abuild/rpmbuild/BUILDROOT/python-django-oscar-3.1-8.8.x86_64/usr/lib/python3.8/site-packages/oscar/test/factories/__init__.py:104: in create_product
[  327s]     setattr(product.attr, code, value)
[  327s] /usr/lib/python3.8/site-packages/django/utils/functional.py:310: in __setattr__
[  327s]     self._setup()
[  327s] /usr/lib/python3.8/site-packages/django/utils/functional.py:419: in _setup
[  327s]     self._wrapped = self._setupfunc()
[  327s] /home/abuild/rpmbuild/BUILDROOT/python-django-oscar-3.1-8.8.x86_64/usr/lib/python3.8/site-packages/oscar/apps/catalogue/abstract_models.py:448: in <lambda>
[  327s]     self.attr = SimpleLazyObject(lambda: ProductAttributesContainer(product=self))
[  327s] /home/abuild/rpmbuild/BUILDROOT/python-django-oscar-3.1-8.8.x86_64/usr/lib/python3.8/site-packages/oscar/apps/catalogue/product_attributes.py:23: in __init__
[  327s]     self.refresh()
[  327s] /home/abuild/rpmbuild/BUILDROOT/python-django-oscar-3.1-8.8.x86_64/usr/lib/python3.8/site-packages/oscar/apps/catalogue/product_attributes.py:26: in refresh
[  327s]     values = self.get_values().select_related('attribute')
[  327s] /home/abuild/rpmbuild/BUILDROOT/python-django-oscar-3.1-8.8.x86_64/usr/lib/python3.8/site-packages/oscar/apps/catalogue/product_attributes.py:51: in get_values
[  327s]     return self.product.get_attribute_values()
[  327s] /home/abuild/rpmbuild/BUILDROOT/python-django-oscar-3.1-8.8.x86_64/usr/lib/python3.8/site-packages/oscar/apps/catalogue/abstract_models.py:671: in get_attribute_values
[  327s]     attribute_values = self.attribute_values.all()
[  327s] /usr/lib/python3.8/site-packages/django/db/models/manager.py:164: in all
[  327s]     return self.get_queryset()
[  327s] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[  327s] 
[  327s] self = <django.db.models.fields.related_descriptors.create_reverse_many_to_one_manager.<locals>.RelatedManager object at 0x7f385b393f40>
[  327s] 
[  327s]     def get_queryset(self):
[  327s]         # Even if this relation is not to pk, we require still pk value.
[  327s]         # The wish is that the instance has been already saved to DB,
[  327s]         # although having a pk value isn't a guarantee of that.
[  327s]         if self.instance.pk is None:
[  327s] >           raise ValueError(
[  327s]                 f"{self.instance.__class__.__name__!r} instance needs to have a "
[  327s]                 f"primary key value before this relationship can be used."
[  327s]             )
[  327s] E           ValueError: 'Product' instance needs to have a primary key value before this relationship can be used.
[  327s] 
[  327s] /usr/lib/python3.8/site-packages/django/db/models/fields/related_descriptors.py:687: ValueError
```

**Comment by jsma** [comment]
As far as I can tell, this test failure can be fixed by something like https://github.com/GreenLightGo/django-oscar/commit/34aa580c81b9faeff4d028a7e77a4ca2f799988a

I'm not sure why this ever worked in older versions of Django, but the code was not saving an actual Basket instance when instantiating for AnonymousUser (with a code comment saying it was unnecessary to save it but no explanation why).

**Comment by jayvdb** [comment]
Applying https://github.com/GreenLightGo/django-oscar/commit/34aa580c81b9faeff4d028a7e77a4ca2f799988a reduces the number of failures from 95 to 36, which are all still a mix of `ValueError: 'Product' instance needs to have a primary key value before this relationship can be used.` and `ValueError: 'Basket' instance needs to have a primary key value before this relationship can be used.`

**Comment by viggo-devries** [comment]
@jsma @jayvdb @solarissmoke Django 4.2 has now been merged to master
Please test if this all works for you now!

**Comment by kaybeudeker** [comment]
Just wondering, is there a (public) roadmap regarding releases? With Django at version 4.2 (LTS), and Python at version 3.11, the latest release of this package feels a bit behind. 

Maybe it's a good idea to use GitHub's Milestones again in this repository to work towards that milestone together and gain some insights?

**Comment by Bastilla123** [comment]
Can anyone tell me how fix:

Applying https://github.com/GreenLightGo/django-oscar/commit/34aa580c81b9faeff4d028a7e77a4ca2f799988a reduces the number of failures from 95 to 36, which are all still a mix of ValueError: 'Product' instance needs to have a primary key value before this relationship can be used. and ValueError: 'Basket' instance needs to have a primary key value before this relationship can be used.

Because this link are not working
